futuredsp/src/firdes.rs

```diff
@@ -1,109 +1,106 @@
 //! Methods for designing FIR filters.
 
 extern crate alloc;
-use crate::windows::FilterWindow;
 use alloc::vec::Vec;
+use num_traits::FromPrimitive;
 
 /// Constructs a lowpass FIR filter with unit gain and cutoff frequency `cutoff` (in cycles/sample)
-/// using the specified window.
+/// using the specified window. The length of the filter equals the length of `window`.
+/// The filter taps are constructed internally as `f64` and then casted to the generic type `T`
+/// using [`num_traits::FromPrimitive::from_f64()`].
 ///
 /// Example usage:
 /// ```
 /// use futuredsp::{firdes, windows};
 ///
 /// let sampling_freq = 10_000;
 /// // 2000 Hz cutoff frequency, rectangular window
-/// let cutoff = 2_000.0 / sampling_freq as f32;
+/// let cutoff = 2_000.0 / sampling_freq as f64;
 /// let num_taps = 65;
-/// let rect_win = windows::RectWindow::new(num_taps);
-/// let taps = firdes::lowpass(num_taps, cutoff, rect_win);
+/// let rect_win = windows::rect(num_taps);
+/// let taps = firdes::lowpass::<f32>(cutoff, rect_win.as_slice());
 /// ```
-pub fn lowpass<T: FilterWindow<f32>>(num_taps: usize, cutoff: f32, window: T) -> Vec<f32> {
-    assert!(num_taps > 0, "num_taps must be greater than 0");
+pub fn lowpass<T: FromPrimitive>(cutoff: f64, window: &[f64]) -> Vec<T> {
     assert!(
         cutoff > 0.0 && cutoff < 1.0 / 2.0,
         "cutoff must be in (0, 1/2)"
     );
-    let mut taps: Vec<f32> = Vec::with_capacity(num_taps);
-    let omega_c = 2.0 * core::f32::consts::PI * cutoff;
-    let alpha = (num_taps - 1) as f32 / 2.0;
-    for n in 0..num_taps {
-        let x = n as f32 - alpha;
-        let tap = match x == 0.0 {
+    let mut taps = window.to_vec();
+    let omega_c = 2.0 * core::f64::consts::PI * cutoff;
+    let alpha = (taps.len() - 1) as f64 / 2.0;
+    for (n, tap) in taps.iter_mut().enumerate() {
+        let x = n as f64 - alpha;
+        let filter_tap = match x == 0.0 {
             true => 1.0,
-            false => (omega_c * x).sin() / (core::f32::consts::PI * x),
+            false => (omega_c * x).sin() / (core::f64::consts::PI * x),
         };
-        taps.push(tap * window.get(n));
+        *tap *= filter_tap;
     }
```

Comment:
If you want, I think this could also be something like:
```
window.iter().enumerate().map(|(n, tap)| {
        let x = n as f64 - alpha;
        let filter_tap = match x == 0.0 {
            true => 1.0,
            false => (omega_c * x).sin() / (core::f64::consts::PI * x),
        };
        tap * filter_tap;
}).map(|x| T::from_f64(*x).unwrap()).collect::<Vec<T>>()
```
to save a heap allocation.

Reply:
Thanks!